# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [9]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [10]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [11]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [12]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [13]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [14]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [15]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [16]:
df.describe()

Season         Daynum          Wteam         Wscore   
count  145289.000000  145289.000000  145289.000000  145289.000000  \
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [17]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [18]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [19]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [20]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [21]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [28]:
# Write your code here
unique_seasons = df['Season'].nunique()
print("Number of unique seasons:", unique_seasons)



Number of unique seasons: 32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [29]:
# Write your code here
most_wins_team = df['Wteam'].value_counts().idxmax()
print("Team with the most wins:", most_wins_team)


Team with the most wins: 1181


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [30]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [31]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [32]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [33]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [34]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [35]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [36]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [37]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [38]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [39]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [40]:
# Write your code here
# Sort by season
df_sorted_by_season = df.sort_values('Season')

# Sort by winning team
df_sorted_by_winning_team = df.sort_values('Wteam')

# Sort by winning score
df_sorted_by_winning_score = df.sort_values('Wscore')

# Select rows from index 100 to 200 and columns for season, winning team, and winning score
selected_rows_columns = df_sorted_by_season.iloc[100:201, [0, 2, 3]]

# Display the results
print("Sorted by Season:")
print(df_sorted_by_season.head())

print("\nSorted by Winning Team:")
print(df_sorted_by_winning_team.head())

print("\nSorted by Winning Score:")
print(df_sorted_by_winning_score.head())

print("\nSelected Rows and Columns:")
print(selected_rows_columns)


Sorted by Season:
      Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0       1985      20   1228      81   1328      64    N      0
2484    1985      99   1275      73   1132      63    H      0
2485    1985      99   1306      62   1171      59    H      0
2486    1985      99   1318      99   1145      91    H      0
2487    1985      99   1337      75   1109      56    H      0

Sorted by Winning Team:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0
136740    2015      62   1101      87   1146      70    H      0
143283    2016      89   1101      80   1249      71    H      0
143476    2016      92   1101      79   1223      72    H      0
141972    2016      61   1101      62   1146      61    A      0

Sorted by Winning Score:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
128582    2013     116   1264      34   1193      31    H      0
69336     2001     126   1206    

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [41]:
# Write your code here
# Find the game with the highest winning score
highest_winning_score_game = df_sorted_by_winning_score.iloc[-1]

# Extract season and winning team from the result
season_highest_score = highest_winning_score_game['Season']
winning_team_highest_score = highest_winning_score_game['Wteam']

# Display the results
print("Season and Winning Team for the Game with the Highest Winning Score:")
print("Season:", season_highest_score)
print("Winning Team:", winning_team_highest_score)


Season and Winning Team for the Game with the Highest Winning Score:
Season: 1991
Winning Team: 1258


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [42]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [43]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [44]:
# Write your code here
# Create a new column 'ScoreDifference'
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Calculate the average 'ScoreDifference'
average_score_difference = df['ScoreDifference'].mean()

# Filter the DataFrame
filtered_df = df[df['ScoreDifference'] > average_score_difference]

# Display the results
print("DataFrame with ScoreDifference greater than Average:")
print(filtered_df.head())


DataFrame with ScoreDifference greater than Average:
    Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference
0     1985      20   1228      81   1328      64    N      0               17
3     1985      25   1165      70   1432      54    H      0               16
6     1985      25   1228      64   1226      44    N      0               20
8     1985      25   1260      98   1133      80    H      0               18
10    1985      25   1307     103   1288      71    H      0               32


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [45]:
# Write your code here
# Find the game with the highest 'ScoreDifference' in the filtered DataFrame
highest_score_difference_game = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]

# Extract season, winning team, and losing team from the result
season_highest_score_difference = highest_score_difference_game['Season']
winning_team_highest_score_difference = highest_score_difference_game['Wteam']
losing_team_highest_score_difference = highest_score_difference_game['Lteam']

# Display the results
print("Season and Teams for the Game with the Highest ScoreDifference:")
print("Season:", season_highest_score_difference)
print("Winning Team:", winning_team_highest_score_difference)
print("Losing Team:", losing_team_highest_score_difference)


Season and Teams for the Game with the Highest ScoreDifference:
Season: 1996
Winning Team: 1409
Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [46]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [47]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       N         3
       A         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [48]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [49]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [50]:
# Write your code here
# Group the DataFrame by season and calculate the average winning score
average_winning_score_by_season = df.groupby('Season')['Wscore'].mean()

# Display the results
print("Average Winning Score for Each Season:")
print(average_winning_score_by_season)


Average Winning Score for Each Season:
Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [51]:
# Write your code here
# Group the DataFrame by winning team and find the maximum winning score
max_winning_score_by_team = df.groupby('Wteam')['Wscore'].max()

# Display the results
print("Maximum Winning Score for Each Team Across All Seasons:")
print(max_winning_score_by_team)



Maximum Winning Score for Each Team Across All Seasons:
Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [52]:
# Write your code here
# Group the DataFrame by both season and winning team, then calculate the average winning score
average_winning_score_by_season_team = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# Find the team with the highest average winning score for each season
team_highest_avg_score_by_season = average_winning_score_by_season_team.groupby('Season').idxmax()

# Display the results
print("Team with the Highest Average Winning Score for Each Season:")
for season, team_idx in team_highest_avg_score_by_season.items():
    team = team_idx[1]
    avg_score = average_winning_score_by_season_team.loc[team_idx]
    print(f"Season {season}: Team {team} with average winning score {avg_score}")


Team with the Highest Average Winning Score for Each Season:
Season 1985: Team 1328 with average winning score 92.8
Season 1986: Team 1109 with average winning score 91.2
Season 1987: Team 1380 with average winning score 95.875
Season 1988: Team 1258 with average winning score 111.75
Season 1989: Team 1258 with average winning score 117.3157894736842
Season 1990: Team 1258 with average winning score 126.34782608695652
Season 1991: Team 1380 with average winning score 112.3125
Season 1992: Team 1380 with average winning score 99.64285714285714
Season 1993: Team 1380 with average winning score 101.875
Season 1994: Team 1380 with average winning score 106.58333333333333
Season 1995: Team 1206 with average winning score 102.83333333333333
Season 1996: Team 1206 with average winning score 103.6
Season 1997: Team 1254 with average winning score 94.14285714285714
Season 1998: Team 1395 with average winning score 101.15384615384616
Season 1999: Team 1317 with average winning score 95.75
Season

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [53]:
# Write your code here
# Group the DataFrame by both season and winning team, then count the number of wins
wins_by_team_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Display the new DataFrame
print("Number of Wins for Each Team in Each Season:")
print(wins_by_team_season.head())


Number of Wins for Each Team in Each Season:
   Season  Wteam  Wins
0    1985   1102     5
1    1985   1103     9
2    1985   1104    21
3    1985   1106    10
4    1985   1108    19


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [54]:
# Write your code here
# Create a DataFrame counting the number of wins for each team in each season
wins_by_team_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Find the team with the most wins for each season
team_with_most_wins_by_season = wins_by_team_season.groupby('Season')['Wins'].idxmax()

# Display the results
print("Team with the Most Wins for Each Season:")
for season, team_idx in team_with_most_wins_by_season.items():
    team = wins_by_team_season.loc[team_idx, 'Wteam']
    wins = wins_by_team_season.loc[team_idx, 'Wins']
    print(f"Season {season}: Team {team} with {wins} wins")


Team with the Most Wins for Each Season:
Season 1985: Team 1385 with 27 wins
Season 1986: Team 1181 with 32 wins
Season 1987: Team 1424 with 33 wins
Season 1988: Team 1112 with 31 wins
Season 1989: Team 1328 with 28 wins
Season 1990: Team 1247 with 29 wins
Season 1991: Team 1116 with 30 wins
Season 1992: Team 1181 with 28 wins
Season 1993: Team 1231 with 28 wins
Season 1994: Team 1163 with 27 wins
Season 1995: Team 1116 with 26 wins
Season 1996: Team 1269 with 31 wins
Season 1997: Team 1242 with 31 wins
Season 1998: Team 1242 with 33 wins
Season 1999: Team 1181 with 32 wins
Season 2000: Team 1409 with 29 wins
Season 2001: Team 1181 with 29 wins
Season 2002: Team 1153 with 30 wins
Season 2003: Team 1166 with 29 wins
Season 2004: Team 1390 with 29 wins
Season 2005: Team 1228 with 32 wins
Season 2006: Team 1181 with 30 wins
Season 2007: Team 1242 with 30 wins
Season 2008: Team 1272 with 33 wins
Season 2009: Team 1272 with 31 wins
Season 2010: Team 1242 with 32 wins
Season 2011: Team 1242 

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [55]:
# Write your code here
# Group the DataFrame by losing team and calculate the average losing score
average_losing_score_by_team = df.groupby('Lteam')['Lscore'].mean()

# Compare with the average winning score from task 3
average_winning_score_by_team = df.groupby('Wteam')['Wscore'].mean()

# Identify teams with a higher average losing score than winning score
teams_with_higher_losing_score = average_losing_score_by_team[average_losing_score_by_team > average_winning_score_by_team]

# Display the results
print("Teams with Higher Average Losing Score than Winning Score:")
print(teams_with_higher_losing_score)


Teams with Higher Average Losing Score than Winning Score:
Series([], Name: Lscore, dtype: float64)


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [56]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [57]:
# Write your code here
# Iterate over the rows and create the 'HighScoringGame' column
df['HighScoringGame'] = ['Yes' if row['Wscore'] > 100 else 'No' for _, row in df.iterrows()]

# Display the DataFrame with the new column
print(df.head())


   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference   
0    1985      20   1228      81   1328      64    N      0               17  \
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   
3    1985      25   1165      70   1432      54    H      0               16   
4    1985      25   1192      86   1447      74    H      0               12   

  HighScoringGame  
0              No  
1              No  
2              No  
3              No  
4              No  


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [58]:
# Write your code here
# Initialize a dictionary to keep track of the number of games for each team
team_games_count = {}

# Iterate over the rows and update the dictionary
for index, row in df.iterrows():
    # Update winning team count
    if row['Wteam'] in team_games_count:
        team_games_count[row['Wteam']] += 1
    else:
        team_games_count[row['Wteam']] = 1
    
    # Update losing team count
    if row['Lteam'] in team_games_count:
        team_games_count[row['Lteam']] += 1
    else:
        team_games_count[row['Lteam']] = 1

# Display the total number of games played by each team
for team, games_count in team_games_count.items():
    print(f"Team {team}: {games_count} games")


Team 1228: 992 games
Team 1328: 968 games
Team 1106: 855 games
Team 1354: 906 games
Team 1112: 981 games
Team 1223: 363 games
Team 1165: 833 games
Team 1432: 69 games
Team 1192: 908 games
Team 1447: 903 games
Team 1218: 931 games
Team 1337: 922 games
Team 1226: 847 games
Team 1242: 993 games
Team 1268: 969 games
Team 1260: 914 games
Team 1133: 949 games
Team 1305: 922 games
Team 1424: 974 games
Team 1307: 969 games
Team 1288: 925 games
Team 1344: 951 games
Team 1438: 952 games
Team 1374: 916 games
Team 1411: 903 games
Team 1412: 962 games
Team 1397: 963 games
Team 1417: 966 games
Team 1225: 880 games
Team 1116: 980 games
Team 1368: 808 games
Team 1120: 936 games
Team 1391: 879 games
Team 1135: 847 games
Team 1306: 898 games
Team 1143: 947 games
Team 1388: 897 games
Team 1153: 970 games
Team 1184: 863 games
Team 1159: 887 games
Team 1171: 829 games
Team 1216: 930 games
Team 1173: 960 games
Team 1134: 200 games
Team 1177: 942 games
Team 1296: 879 games
Team 1193: 942 games
Team 1265: 934

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [59]:
# Write your code here
# Initialize a dictionary to keep track of the highest score difference for each season
highest_score_difference_by_season = {}

# Iterate over the rows and update the dictionary
for index, row in df.iterrows():
    season = row['Season']
    score_difference = row['ScoreDifference']
    
    # Update the highest score difference for each season
    if season in highest_score_difference_by_season:
        current_max_difference = highest_score_difference_by_season[season]
        if score_difference > current_max_difference:
            highest_score_difference_by_season[season] = score_difference
    else:
        highest_score_difference_by_season[season] = score_difference

# Display the game with the highest score difference for each season
for season, max_difference in highest_score_difference_by_season.items():
    game_index = df[(df['Season'] == season) & (df['ScoreDifference'] == max_difference)].index[0]
    print(f"Season {season}: Game with the highest score difference is at index {game_index} with difference {max_difference}")



Season 1985: Game with the highest score difference is at index 236 with difference 60
Season 1986: Game with the highest score difference is at index 4731 with difference 84
Season 1987: Game with the highest score difference is at index 8240 with difference 73
Season 1988: Game with the highest score difference is at index 12046 with difference 68
Season 1989: Game with the highest score difference is at index 16677 with difference 70
Season 1990: Game with the highest score difference is at index 19502 with difference 76
Season 1991: Game with the highest score difference is at index 25161 with difference 68
Season 1992: Game with the highest score difference is at index 27997 with difference 82
Season 1993: Game with the highest score difference is at index 33858 with difference 81
Season 1994: Game with the highest score difference is at index 36404 with difference 69
Season 1995: Game with the highest score difference is at index 39858 with difference 74
Season 1996: Game with th

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [60]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [61]:
# Write your code here
# Calculate the total number of games played by each team using vectorized operations
wins_count = df['Wteam'].value_counts()
losses_count = df['Lteam'].value_counts()

# Add the two Series together to get the total number of games played by each team
total_games_by_team = wins_count.add(losses_count, fill_value=0)

# Display the results
print("Total Number of Games Played by Each Team:")
print(total_games_by_team.head())


Total Number of Games Played by Each Team:
1101     76
1102    840
1103    910
1104    975
1105    447
Name: count, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [62]:
# Write your code here
# Create a new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Find the game with the highest score difference for each season
max_score_difference_by_season = df.groupby('Season')['ScoreDifference'].idxmax()

# Display the results
print("Game with the Highest Score Difference for Each Season:")
for season, game_index in max_score_difference_by_season.items():
    print(f"Season {season}: Game with the highest score difference is at index {game_index} with difference {df.loc[game_index, 'ScoreDifference']}")


Game with the Highest Score Difference for Each Season:
Season 1985: Game with the highest score difference is at index 236 with difference 60
Season 1986: Game with the highest score difference is at index 4731 with difference 84
Season 1987: Game with the highest score difference is at index 8240 with difference 73
Season 1988: Game with the highest score difference is at index 12046 with difference 68
Season 1989: Game with the highest score difference is at index 16677 with difference 70
Season 1990: Game with the highest score difference is at index 19502 with difference 76
Season 1991: Game with the highest score difference is at index 25161 with difference 68
Season 1992: Game with the highest score difference is at index 27997 with difference 82
Season 1993: Game with the highest score difference is at index 33858 with difference 81
Season 1994: Game with the highest score difference is at index 36404 with difference 69
Season 1995: Game with the highest score difference is at 

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [63]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [64]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [65]:
type(df['Wscore'])

pandas.core.series.Series

In [66]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [69]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference   
0    1985      20   1228      81   1328      64    N      0               17  \
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [68]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference   
0    1985      20   1228      81   1328      64    N      0               17  \
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [67]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y